File: ConsoleProject/RaceConditionTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Diagnostics;

namespace ConsoleProject
{
	public class RaceConditionTest
	{

		//http://www.gotw.ca/publications/concurrency-ddj.htm
		public static void Test()
		{
			Stopwatch watch = new Stopwatch();
			watch.Start();
			List<int> list = new List<int>();
			int count = Environment.ProcessorCount;

			for(int i = 0; i < Environment.ProcessorCount; i++)
			{
				Thread.Sleep(1000);
				list.Add(i);
			};

			Console.WriteLine($"The total count should be {count}. The list count is {list.Count}. It took {watch.ElapsedMilliseconds} ms.");
		}


		public static void Test1()
		{
			Stopwatch watch = new Stopwatch();
			watch.Start();
			List<int> list = new List<int>();
			int count = 8;

			Parallel.For(0, count, new ParallelOptions() { MaxDegreeOfParallelism = Environment.ProcessorCount }, (i) =>
			{
				Thread.Sleep(1000);
				list.Add(i);
			});

			Console.WriteLine($"The total count should be {count}. The list count is {list.Count}. It took {watch.ElapsedMilliseconds} ms.");
		}
	}
}

File: FConsoleProject/RaceConditionTest.fs
namespace FConsole
module RaceConditionTest = 
    open System
    open System.Diagnostics
    open System.Threading

    /// Add numbers to list
    let addUpTo count =
        let wait value = async { 
            Thread.Sleep 1000
            return value
        }
        seq { for i in 0 .. (count - 1) -> wait i } |> List.ofSeq |> Async.Parallel |> Async.RunSynchronously

    let test () = 
        let watch = new Stopwatch()
        watch.Start()
        let count = Environment.ProcessorCount;
        //let count = 1000000;

        let list = addUpTo count 
        printfn "The total count should be %d. The list count is %d. It took %d ms." count list.Length watch.ElapsedMilliseconds
        